### PR TITLE
Add `no_std` support 

### PR DESCRIPTION
**Pull request by althonos**
Hi @srijs,

very nice implementation you have here ! I added feature-based support for `no_std` using the `raw-cpuid` crate to provide CPU features information at runtime.

Not much to say, this PR is quite straightforward and should not change much !

### CODE REVIEW

**Comment by srijs** [comment]
Hi, thanks for your PR! Having a `no_std` feature is a great idea, and am interested in getting this merged.

With regards to your approach, I was reading https://github.com/rust-lang-nursery/stdsimd/issues/464 earlier, in which the following was pointed out by @gnzlbg:

> Just keep in mind that a cpuid bit being set does not always imply that a feature can be used correctly (that is, without invoking undefined behavior). For some features like some of the SIMD related ones whether a feature can be used correctly will also depend on whether OS support is required, and if that's the case, on whether it is properly enabled.

Admittedly, I don't fully understand all the implications of this, so I have to trust their expertise as one of the maintainers of the stdsimd crate.

At this point, unless someone can help me understand this better, I think I might be more comfortable with having a `no_std` mode that either always falls back to the baseline implementation, or does a static feature check via `cfg!(target_feature = ...)`.

**Comment by gnzlbg** [comment]
> At this point, unless someone can help me understand this better,

The TL;DR is that, often, operating system have to be aware of which registers user-space applications are actually using to be able to preserve them across context switches - these registers have to be saved to memory, then another application runs for a while, and before the previous application continues its registers are restored from memory. 

If the OS "says" that user-space applications are not allowed to use FPU/MMX/SSE/AVX/AVX512/... registers, then it does not have to copy the contents of those registers to memory and back while doing a context switch, reducing memory usage, making context switches faster, etc.

The approach pursued here will tell you if the CPU supports these features, but it won't tell you whether the operating system does. If it doesn't, those registers will contain garbage after a context switch, and the behavior is undefined.

File: Cargo.toml [hunk 1]
```diff
@@ -13,12 +13,17 @@ keywords = ["checksum", "crc", "crc32", "simd", "fast"]
...
+raw-cpuid = { version = "6.1", optional = true }
 
 [dev-dependencies]
 bencher = "0.1"
 quickcheck = { version = "0.6", default-features = false }
 rand = "0.4"
 
+[features]
+default = []
+no_std = ["raw-cpuid"]
```
Comment (gnzlbg) [inline Cargo.toml]: The general API guideline is that features should be additive. Pretty much every official / semi-official rust-lang / rust-lang-nursery library uses `use_std` as the feature, which one can enable by default.

**Comment by srijs** [comment]
Thanks @gnzlbg, I appreciate you taking the time to explain this in such detail. This makes total sense to me, and convinces me further that we don't want to use a cpuid crate here, but instead rely on compile-time detection only.

Also thanks for your comment about `no_std` vs `use_std`, we should definitely follow the official standard here.

File: Cargo.toml (hunk 1 above)
Comment (althonos) [reply Cargo.toml]: I know that, but mostly because they have more dependencies when `std` is enabled, which is not the case here. Do you know the proper way to have a `use_std` optional feature, enable by default, which when disabled makes the `raw_cpuid` a requirement ?

**Comment by althonos** [comment]
What about using this crate in a true bare-metal context, e.g. building an OS ? In this case you wouldn't have to worry about context switches. I understand the issues with targets that have an OS, but for targets that have none, this could still be useful. 

Here are the changes I can provide:
- use an additive style feature (i.e. `use_std`) to enable building against `std`, enabled by default
- add a `cpuid` feature that enables the use of `raw-cpuid`, but only in conjunction with the `target_os = "none"` feature, disabled by default

This way, enabling the `cpuid` feature requires manual intervention even in `no_std` mode AND won't do anything if you're building to a system with an OS.

File: Cargo.toml (hunk 1 above)
Comment (gnzlbg) [reply Cargo.toml]: > Do you know the proper way to have a use_std optional feature, enable by default, which when disabled makes the raw_cpuid a requirement ?

Something like

```toml
[target.`cfg(not(feature = "use_std"))`.dependencies.raw_cpuid]
version = "6.1"
```

?

**Comment by gnzlbg** [comment]
> What about using this crate in a true bare-metal context, e.g. building an OS ?

Many OSes, e.g., [including Linux](https://stackoverflow.com/questions/46677676/why-are-simd-instructions-not-used-in-kernel), do not allow using any FP/SIMD/... registers in kernel space for a variety of reasons. 

The answer still is that it depends on whether the OS being written allows using these or not.

File: Cargo.toml (hunk 1 above)
Comment (gnzlbg) [reply Cargo.toml]: Damn, giving these a try it did not work, and I ended up landing on the same cargo bug that you mentioned 😆

**Comment by althonos** [comment]
Fine by me ! I'll remove the `raw-cpuid` implementation, that should resolve most questions here.

**Comment by gnzlbg** [comment]
One thing that one could do is expose the APIs here "unsafely" (e.g. as `unsafe fn` somehow), and use `#[target_feature(...)]` to unconditionally compile these with the features enabled.

One can then expose "safe" APIs that use `cfg(target_feature = "...")`, `is_..._feature_detected` when `libstd` is available, etc. 

That way, a `#[no_std]` user could still use the implementation that requires the features, either statically, or dynamically, but to do so dynamically its up to the user to do the run-time feature detection.

File: README.md [hunk 2]
```diff
@@ -47,6 +47,13 @@ On top of that, every commit is tested using an address sanitizer in CI to catch
 
 Even though neither fuzzing not sanitization has revealed any safety bugs yet, please don't hesitate to file an issue if you run into any crashes or other unexpected behaviour.
 
+## `no_std`
```
Comment (srijs) [inline README.md]: It looks like this paragraph needs some changes to reflect the new feature flag approach.

File: src/specialized/pclmulqdq.rs [hunk 3]
```diff
@@ -51,6 +57,7 @@ const K6: i64 = 0x1db710640;
 const P_X: i64 = 0x1DB710641;
 const U_PRIME: i64 = 0x1F7011641;
 
+#[cfg(feature = "use_std")]
```
Comment (srijs) [inline src/specialized/pclmulqdq.rs]: This looks like a typo, should this be `#[cfg(feature = "std")]` instead? (same on the other definition below)

**Comment by srijs** [comment]
I've manually cherry-picked your changes on master to resolve the merge conflict. Thanks again for your PR! 💯
